pkg/plugins/common/postgres/pgx_listener.go

```diff
@@ -0,0 +1,209 @@
+	ctx := context.Background()
+	connectionString, err := config.ConnectionString()
+	if err != nil {
+		return nil, err
+	}
+	db, err := pgxpool.New(context.Background(), connectionString)
+	if err != nil {
+		return nil, err
+	}
+	l := &PgxListener{
+		notificationsCh: make(chan *Notification, 32),
+		logger:          logger,
+		ctx:             ctx,
+		channel:         "resource_events",
+		db:              db,
+	}
+
+	err = l.connect(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	l.Start()
+
+	return l, nil
+}
+
+func (l *PgxListener) Start() {
+	l.logger.Info("started")
+
+	l.mx.Lock()
+	defer l.mx.Unlock()
+
+	if l.running {
+		return
+	}
+
+	ctx, cancel := context.WithCancel(l.ctx)
+	l.stopFn = cancel
+	l.stoppedCh = make(chan struct{}, 1)
+
+	go l.run(ctx)
+
+	l.running = true
+}
+
+func (l *PgxListener) Close() error {
+	l.stop()
+	close(l.notificationsCh)
+	return nil
+}
+
+func (l *PgxListener) run(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			l.stoppedCh <- struct{}{}
+			return
+		default:
+		}
+		err := l.handleNotifications(ctx)
+		if errors.Is(err, context.Canceled) {
+			err = nil
+		}
+		if err != nil {
+			l.logger.Error(err, "error handling notification")
+		}
+	}
+}
+
+// stop will end all current connections and stop reconnecting.
+func (l *PgxListener) stop() {
+	l.logger.Info("stopped")
+	l.mx.Lock()
+	defer l.mx.Unlock()
+
+	if !l.running {
+		return
+	}
+
+	l.stopFn()
+	<-l.stoppedCh
+
+	l.running = false
+}
+
+func (l *PgxListener) handleNotifications(ctx context.Context) error {
+	defer l.disconnect()
+	t := time.NewTicker(3 * time.Second)
+	defer t.Stop()
+	for {
+		err := l.connect(ctx)
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+		if err != nil {
+			l.logger.Error(err, "error connecting")
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-t.C:
+				continue
+			}
+		}
+		break
+	}
+
+	notification, err := l.conn.Conn().WaitForNotification(ctx)
+	if err != nil {
+		return err
+	}
+
+	l.logger.Info("event happened", "event", notification)
+
+	select {
+	// Writing to channel
+	case l.notificationsCh <- toBareNotification(notification):
+	default:
+	}
+
+	return nil
+}
+
+func toBareNotification(notification *pgconn.Notification) *Notification {
+	return &Notification{
+		Payload: notification.Payload,
+	}
+}
+
+func (l *PgxListener) Notify() chan *Notification { return l.notificationsCh }
+
+func (l *PgxListener) disconnect() {
+	if l.conn == nil {
+		return
+	}
+	l.conn.Release()
+	l.conn = nil
+}
+
+func (l *PgxListener) connect(ctx context.Context) error {
+	if l.conn != nil {
+		return nil
+	}
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+	conn, err := l.db.Acquire(ctx)
+	if err != nil {
+		return errors.Wrap(err, "get connection")
+	}
+
+	l.conn = conn
+
+	select {
+	case <-ctx.Done():
+		l.disconnect()
+		return ctx.Err()
+	default:
+	}
```

Comment:
same question, could we just rely on `conn.Exec(ctx` ?

Reply:
👍